### PR TITLE
Update main

### PR DESCRIPTION
**Pull request by Bklzn**
Merge all dev changes into main

### CODE REVIEW

File: backend/src/app/web_api.py [hunk 1]
```diff
@@ -228,14 +397,299 @@ async def get_paginated_transactions(
...
 
 
+@app.get("/api/statistics/transactions_cost", response_class=JSONResponse)
+async def get_transaction_usage_statistics_over_time(
+    request: Request,
+    ctx: Annotated[TransactionContext, Depends(get_transaction_context)],
+    project_id: str,
+    date_from: datetime | str | None = None,
+    date_to: datetime | str | None = None,
+    period: str | None = "daily",
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py]: I've already corrected it

File: backend/src/app/web_api.py [hunk 2]
```diff
@@ -228,14 +397,299 @@ async def get_paginated_transactions(
...
+    total cost calculated based on cumulative values for the best possible representation of costs over time.\n
+
+    :param request: The FastAPI Request object (automatically applied).\n
+    :param ctx: The transaction context, providing access to dependencies (automatically applied).\n
+    :param project_id: The unique identifier of the project.\n
+    :param date_from: Starting point of the time interval (optional - when empty, then the scope is counted from the
+        beginning of the project's existence).\n
+    :param date_to: End point of the time interval (optional - when empty, then the interval is counted up to the
+        present time).\n
+    :param period: The time period for grouping statistics - can be yearly, monthly, weekly, daily, hourly or minutely.
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py]: Now it is an enum field with values:
- year -> year
- month -> month
- week -> week
- day -> day
- hour -> hour
- minutes -> 5minutes (Due to the inability to name a field starting with a number)

File: backend/src/app/web_api.py [hunk 3]
```diff
@@ -228,14 +397,299 @@ async def get_paginated_transactions(
...
+                total_transactions=1,
+            )
+            for transaction in transactions
+        ]
+        stats = utils.status_counter_for_transactions(
+            transactions, period, date_from, date_to
+        )
+
+        return stats
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py]: The structure of the middleware has been minimally changed due to summer's built-in support for middlewares, but this has been resolved as indicated.

File: backend/src/app/web_api.py [hunk 4]
```diff
@@ -228,14 +397,299 @@ async def get_paginated_transactions(
...
+                date=transaction.response_time,
+                latency=(
+                    transaction.response_time - transaction.request_time
+                ).total_seconds(),
+                generation_speed=transaction.generation_speed,
+                total_transactions=1,
+            )
+            for transaction in transactions
+        ]
+        stats = utils.latency_counter_for_transactions(
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py]: Name changed to speed_counter_for_transactions

[the remark this answers is not in the document]
Comment (ksopyla) [reply backend/src/app/web_api.py]: @sliwaszymon This is a slight change from my side; I think a better name would be `generation_speed_mean` or 'generation_speed_average_by_period`. This better reveals the character of the method. 
Counter means that we count something but instead we aggregate average gen speed.

File: backend/src/app/web_api.py [hunk 5]
```diff
@@ -271,3 +692,34 @@
...
+            repo.add(transaction)
+        time_stop = datetime.now(tz=timezone.utc)
+
+        return {
+            "status_code": 200,
+            "message": f"{count} transactions added in {(time_stop-time_start).total_seconds()} seconds.",
+        }
+
+    except Exception as e:
+        return {"status_code": 500, "message": str(e)}
```
[the remark this answers is not in the document]
Comment (ksopyla) [reply backend/src/app/web_api.py:748]: @sliwaszymon Do we need an explicit exception message return here? Can it not be handled by exception middleware?

File: backend/src/app/web_api.py [hunk 6]
```diff
@@ -616,60 +609,58 @@ async def get_transaction_latency_statistics_over_time(
...
-                date=transaction.response_time,
-                latency=(
-                    transaction.response_time - transaction.request_time
-                ).total_seconds(),
-                generation_speed=transaction.generation_speed,
-                total_transactions=1,
-            )
-            for transaction in transactions
-        ]
-        stats = utils.speed_counter_for_transactions(
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py]: This aggregates by provider and model and then for the time series counts the average latency and generation speed in tokens per second. Then any ideas? Last times it was sth. like generation speed counter and it didn't fit either.

File: backend/src/app/web_api.py [hunk 7]
```diff
@@ -616,60 +609,58 @@ async def get_transaction_latency_statistics_over_time(
...
+        date_from=date_from,
+        date_to=date_to,
+    )
+    if count == 0:
+        return []
 
-        transactions = ctx.call(
-            get_list_of_filtered_transactions,
+    transactions = ctx.call(
+        get_list_of_filtered_transactions,
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py:637]: just above ;) first I check the number of transactions, if it is equal to 0 I immediately return an empty list

File: backend/src/app/web_api.py [hunk 8]
```diff
@@ -427,87 +427,83 @@ async def get_transaction_usage_statistics_over_time(
...
+        date_from=date_from,
+        date_to=date_to,
+    )
+    if count == 0:
+        return []
 
-        count = ctx.call(
-            count_transactions,
+    transactions = ctx.call(
+        get_list_of_filtered_transactions,
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py:454]: lines 444-451

File: backend/src/app/web_api.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ksopyla) [reply backend/src/app/web_api.py]: @sliwaszymon Here, I refer to the word "counter"; it doesn't reflect what the function does; the function does not count but computes the average. 
Last time it was  "latency_counter_for_transactions"

File: backend/src/app/web_api.py [hunk 9]
```diff
@@ -536,60 +532,57 @@ async def get_transaction_status_statistics_over_time(
...
-            if len(date_from) == 10:
-                date_from = datetime.fromisoformat(str(date_from) + "T00:00:00")
-            else:
-                date_from = datetime.fromisoformat(date_from)
-        if isinstance(date_to, str):
-            if len(date_to) == 10:
-                date_to = datetime.fromisoformat(date_to + "T00:00:00")
-            else:
-                date_to = datetime.fromisoformat(date_to)
+    if isinstance(date_from, str):
```
[the remark this answers is not in the document]
Comment (sliwaszymon) [reply backend/src/app/web_api.py]: Got it! Now its separated to util function

File: backend/src/app/web_api.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ksopyla) [reply backend/src/app/web_api.py:597]: @sliwaszymon I'm not referring to the zero transactions in a specified period; this is ok, but to a situation where in a specified period we have a few transactions, but some of them failed (status code <> 200). 
Lets look at the example, suppose we have 3 transactions: 
1. tr1 - input_tokens:1234, output_tokens: 550, latency (response_time - request_time) = 5.5 s, gen_speed=100 status_code: 200
2. tr2 - input_tokens:1234, output_tokens: 100, latency (response_time - request_time) =2s, gen_speed=50 status_code: 200
3. tr1 - input_tokens:1234, output_tokens: **error**, latency (response_time - request_time) = 1.34 s, **gen_speed=??** status_code: **404**

And now we could compute this in two ways: 
1. We will aggregate gen_speed of all of the transitions, so (100+50+ 0?)/3 = 50
2. we will aggregate gen_speed of transactions which generate something (100+50)/2 = 75
In my opinion option 2 makes more sense.

**Pull request merged**